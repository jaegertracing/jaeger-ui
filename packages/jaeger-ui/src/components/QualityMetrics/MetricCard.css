/*
Copyright (c) 2020 The Jaeger Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.MetricCard {
  border-radius: 7px;
  border: 1px rgba(0, 0, 0, 0.5) solid;
  display: flex;
  margin: 5px 5%;
}

.MetricCard--Body {
  border-left: solid 1px rgba(0, 0, 0, 0.3);
  box-shadow: -3px 0 3px rgba(0, 0, 0, 0.1);
  flex-grow: 1;
  padding: 5px;
  text-align: center;
}

.MetricCard--CircularProgressbarWrapper {
  flex-shrink: 0;
  max-width: 180px;
  padding: 5px;
  width: 25%;
}

.MetricCard--CountsWrapper {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
}

.MetricCard--Description {
  margin: 0 auto;
  width: 60%;
}

.MetricCard--Details {
  text-align: left;
  padding: 0.3em;
}

.MetricCard--TitleHeader {
  border-bottom: solid 1px #ddd;
  box-shadow: 0px 5px 5px -5px rgba(0, 0, 0, 0.3);
  font-size: 1.5em;
}
