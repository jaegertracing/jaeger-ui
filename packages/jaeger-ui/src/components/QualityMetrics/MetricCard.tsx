// Copyright (c) 2020 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Tooltip } from 'antd';

import CircularProgressbar from '../common/CircularProgressbar';
import NewWindowIcon from '../common/NewWindowIcon';
import DetailsCard from '../common/DetailsCard';
import CountCard from './CountCard';

import { TQualityMetrics } from './types';

import './MetricCard.css';

export type TProps = {
  metric: TQualityMetrics['metrics'][0];
};

const dividToFixedFloorPercentage = (pass: number, fail: number) => {
  const str = `${(pass / (pass + fail)) * 100}.0`;
  return `${str.substring(0, str.indexOf('.') + 2)}%`;
};

export default class MetricCard extends React.PureComponent<TProps> {
  render() {
    const {
      metric: {
        name,
        description,
        metricDocumentationLink,
        passCount,
        passExamples,
        failureCount,
        failureExamples,
        exemptionCount,
        exemptionExamples,
        details,
      },
    } = this.props;
    return (
      <div className="MetricCard">
        <div className="MetricCard--CircularProgressbarWrapper">
          <CircularProgressbar
            backgroundHue={passCount === 0 ? 0 : undefined}
            decorationHue={passCount === 0 ? 0 : 120}
            maxValue={passCount + failureCount}
            text={dividToFixedFloorPercentage(passCount, failureCount)}
            value={passCount}
          />
        </div>
        <div className="MetricCard--Body">
          <span className="MetricCard--TitleHeader">
            {name}{' '}
            <Tooltip arrowPointAtCenter title="Metric Documentation">
              <a href={metricDocumentationLink} target="_blank" rel="noreferrer noopener">
                <NewWindowIcon />
              </a>
            </Tooltip>
          </span>
          <p className="MetricCard--Description">{description}</p>
          <div className="MetricCard--CountsWrapper">
            <CountCard count={passCount} examples={passExamples} title="Passing" />
            <CountCard count={failureCount} examples={failureExamples} title="Failing" />
            <CountCard count={exemptionCount} examples={exemptionExamples} title="Exemptions" />
          </div>
          {details &&
            details.map(
              detail =>
                Boolean(detail.rows && detail.rows.length) && (
                  <DetailsCard
                    key={`${detail.description}${JSON.stringify(detail.columns)}`}
                    className="MetricCard--Details"
                    collapsible
                    columnDefs={detail.columns}
                    description={detail.description}
                    details={detail.rows}
                    header={detail.header || 'Details'}
                  />
                )
            )}
        </div>
      </div>
    );
  }
}
