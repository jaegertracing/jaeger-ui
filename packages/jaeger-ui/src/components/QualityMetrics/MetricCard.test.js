// Copyright (c) 2020 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { shallow } from 'enzyme';

import MetricCard from './MetricCard';

describe('MetricCard', () => {
  const metric = {
    name: 'Metric Name',
    description: 'Metric Description',
    metricDocumentationLink: 'metric.documentation.link',
    passCount: 108,
    passExamples: ['foo'],
    failureCount: 255,
    failureExamples: ['bar'],
    exemptionCount: 42,
    exemptionExamples: ['baz'],
  };
  const details = [
    {
      columns: ['col0', 'col1'],
      description: 'Details[0] Description',
    },
    {
      columns: ['col2', 'col3'],
      description: 'Details[1] Description',
      rows: [],
    },
    {
      columns: ['col4', 'col5'],
      description: 'Details[2] Description',
      rows: [
        {
          col4: 'value for fourth column',
          col5: 'value for fifth column',
        },
      ],
    },
    {
      columns: ['col6', 'col7'],
      description: 'Details[3] Description',
      header: 'Details[3] Header',
      rows: [
        {
          col6: 'value for sixth column',
          col7: 'value for seventh column',
        },
      ],
    },
  ];

  it('renders as expected without details', () => {
    expect(shallow(<MetricCard metric={metric} />)).toMatchSnapshot();
    expect(
      shallow(
        <MetricCard
          metric={{
            ...metric,
            details: details.slice(0, 2),
          }}
        />
      )
    ).toMatchSnapshot();
  });

  it('renders as expected with details', () => {
    expect(
      shallow(
        <MetricCard
          metric={{
            ...metric,
            details,
          }}
        />
      )
    ).toMatchSnapshot();
  });

  it('renders as expected when passCount is zero', () => {
    expect(
      shallow(
        <MetricCard
          metric={{
            ...metric,
            passCount: 0,
          }}
        />
      )
    ).toMatchSnapshot();
  });
});
