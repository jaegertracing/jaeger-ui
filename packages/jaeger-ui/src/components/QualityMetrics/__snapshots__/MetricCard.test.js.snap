// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetricCard renders as expected when passCount is zero 1`] = `
<div
  className="MetricCard"
>
  <div
    className="MetricCard--CircularProgressbarWrapper"
  >
    <CircularProgressbar
      backgroundHue={0}
      decorationHue={0}
      maxValue={255}
      text="0.0%"
      value={0}
    />
  </div>
  <div
    className="MetricCard--Body"
  >
    <span
      className="MetricCard--TitleHeader"
    >
      Metric Name
       
      <Tooltip
        arrowPointAtCenter={true}
        autoAdjustOverflow={true}
        mouseEnterDelay={0.1}
        mouseLeaveDelay={0.1}
        placement="top"
        title="Metric Documentation"
        transitionName="zoom-big-fast"
      >
        <a
          href="metric.documentation.link"
          rel="noreferrer noopener"
          target="_blank"
        >
          <NewWindowIcon
            isLarge={false}
          />
        </a>
      </Tooltip>
    </span>
    <p
      className="MetricCard--Description"
    >
      Metric Description
    </p>
    <div
      className="MetricCard--CountsWrapper"
    >
      <ScoreCard
        count={0}
        examples={
          Array [
            "foo",
          ]
        }
        title="Passing"
      />
      <ScoreCard
        count={255}
        examples={
          Array [
            "bar",
          ]
        }
        title="Failing"
      />
      <ScoreCard
        count={42}
        examples={
          Array [
            "baz",
          ]
        }
        title="Exemptions"
      />
    </div>
  </div>
</div>
`;

exports[`MetricCard renders as expected with details 1`] = `
<div
  className="MetricCard"
>
  <div
    className="MetricCard--CircularProgressbarWrapper"
  >
    <CircularProgressbar
      decorationHue={120}
      maxValue={363}
      text="29.7%"
      value={108}
    />
  </div>
  <div
    className="MetricCard--Body"
  >
    <span
      className="MetricCard--TitleHeader"
    >
      Metric Name
       
      <Tooltip
        arrowPointAtCenter={true}
        autoAdjustOverflow={true}
        mouseEnterDelay={0.1}
        mouseLeaveDelay={0.1}
        placement="top"
        title="Metric Documentation"
        transitionName="zoom-big-fast"
      >
        <a
          href="metric.documentation.link"
          rel="noreferrer noopener"
          target="_blank"
        >
          <NewWindowIcon
            isLarge={false}
          />
        </a>
      </Tooltip>
    </span>
    <p
      className="MetricCard--Description"
    >
      Metric Description
    </p>
    <div
      className="MetricCard--CountsWrapper"
    >
      <ScoreCard
        count={108}
        examples={
          Array [
            "foo",
          ]
        }
        title="Passing"
      />
      <ScoreCard
        count={255}
        examples={
          Array [
            "bar",
          ]
        }
        title="Failing"
      />
      <ScoreCard
        count={42}
        examples={
          Array [
            "baz",
          ]
        }
        title="Exemptions"
      />
    </div>
    <DetailsCard
      className="MetricCard--Details"
      collapsible={true}
      columnDefs={
        Array [
          "col4",
          "col5",
        ]
      }
      description="Details[2] Description"
      details={
        Array [
          Object {
            "col4": "value for fourth column",
            "col5": "value for fifth column",
          },
        ]
      }
      header="Details"
      key="Details[2] Description[\\"col4\\",\\"col5\\"]"
    />
    <DetailsCard
      className="MetricCard--Details"
      collapsible={true}
      columnDefs={
        Array [
          "col6",
          "col7",
        ]
      }
      description="Details[3] Description"
      details={
        Array [
          Object {
            "col6": "value for sixth column",
            "col7": "value for seventh column",
          },
        ]
      }
      header="Details[3] Header"
      key="Details[3] Description[\\"col6\\",\\"col7\\"]"
    />
  </div>
</div>
`;

exports[`MetricCard renders as expected without details 1`] = `
<div
  className="MetricCard"
>
  <div
    className="MetricCard--CircularProgressbarWrapper"
  >
    <CircularProgressbar
      decorationHue={120}
      maxValue={363}
      text="29.7%"
      value={108}
    />
  </div>
  <div
    className="MetricCard--Body"
  >
    <span
      className="MetricCard--TitleHeader"
    >
      Metric Name
       
      <Tooltip
        arrowPointAtCenter={true}
        autoAdjustOverflow={true}
        mouseEnterDelay={0.1}
        mouseLeaveDelay={0.1}
        placement="top"
        title="Metric Documentation"
        transitionName="zoom-big-fast"
      >
        <a
          href="metric.documentation.link"
          rel="noreferrer noopener"
          target="_blank"
        >
          <NewWindowIcon
            isLarge={false}
          />
        </a>
      </Tooltip>
    </span>
    <p
      className="MetricCard--Description"
    >
      Metric Description
    </p>
    <div
      className="MetricCard--CountsWrapper"
    >
      <ScoreCard
        count={108}
        examples={
          Array [
            "foo",
          ]
        }
        title="Passing"
      />
      <ScoreCard
        count={255}
        examples={
          Array [
            "bar",
          ]
        }
        title="Failing"
      />
      <ScoreCard
        count={42}
        examples={
          Array [
            "baz",
          ]
        }
        title="Exemptions"
      />
    </div>
  </div>
</div>
`;

exports[`MetricCard renders as expected without details 2`] = `
<div
  className="MetricCard"
>
  <div
    className="MetricCard--CircularProgressbarWrapper"
  >
    <CircularProgressbar
      decorationHue={120}
      maxValue={363}
      text="29.7%"
      value={108}
    />
  </div>
  <div
    className="MetricCard--Body"
  >
    <span
      className="MetricCard--TitleHeader"
    >
      Metric Name
       
      <Tooltip
        arrowPointAtCenter={true}
        autoAdjustOverflow={true}
        mouseEnterDelay={0.1}
        mouseLeaveDelay={0.1}
        placement="top"
        title="Metric Documentation"
        transitionName="zoom-big-fast"
      >
        <a
          href="metric.documentation.link"
          rel="noreferrer noopener"
          target="_blank"
        >
          <NewWindowIcon
            isLarge={false}
          />
        </a>
      </Tooltip>
    </span>
    <p
      className="MetricCard--Description"
    >
      Metric Description
    </p>
    <div
      className="MetricCard--CountsWrapper"
    >
      <ScoreCard
        count={108}
        examples={
          Array [
            "foo",
          ]
        }
        title="Passing"
      />
      <ScoreCard
        count={255}
        examples={
          Array [
            "bar",
          ]
        }
        title="Failing"
      />
      <ScoreCard
        count={42}
        examples={
          Array [
            "baz",
          ]
        }
        title="Exemptions"
      />
    </div>
  </div>
</div>
`;
